source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -208,6 +215,10 @@ void DnsResolverImpl::AddrInfoPendingResolution::onAresGetAddrInfoCallback(
 
     ASSERT(addrinfo != nullptr);
     ares_freeaddrinfo(addrinfo);
+  } else if (accept_nodata_ && isResponseWithNoRecords(status) && !dual_resolution_) {
```

Comment:
Do we need a !dual_resolution check here?  It looks to me it works even without this check, right?

Reply:
I wanted to make sure if AUTO / V4_PREFERRED is used then we'll only accept nodata response for the second/last attempt (i.e. if the first request returns on records, I still want to give the second request targeting a different record type a chance), please let me know if I'm understanding it incorrectly.

Reply:
For dual_resolution case, even you set the pending_response_.status_ = ResolutionStatus::Success here, it still go ahead send the 2nd request for a different type:  https://github.com/envoyproxy/envoy/blob/a198e966f15ca1ca81258de6762bb1205c4e7a23/source/extensions/network/dns_resolver/cares/dns_impl.cc#L235.

I guess the real questions is that : if the 1st request returns ARES_ENODATA, and the 2nd request returns a wrong status, like ARES_ECONNREFUSED, which status you want to set it for pending_response_.status_?  IIUC, it should be ResolutionStatus::Success since that's a "SUCCESS" and the 2nd request failure can be ignored in this case.     However, if you add the !dual_resolution check here, the code will not set the pending_response_.status_  to ResolutionStatus::Success in such situation.

Reply:
Good point, I think you're right, I originally put a `completed_ = true` in this block and added this check to make sure it won't block the second request, then I realized that another block already set the completed_ for second request (line #165) so I removed it from this block but somehow got the incorrect impression that entering this block will suppress the second request. Your proposal is more aligned with the how the code is handling `ARES_SUCCESS` (i.e. as long as one request returns `ARES_SUCCESS` the final result is a "SUCCESS"), thanks for pointing it out!